tools/net-common.in

```diff
@@ -63,6 +63,41 @@ function fix_mac {
   ip link set $INTERFACE address $FIXED_MAC || true
 }
 
+function bridge_vlan_add {
+  local VID="${1}"
+  shift
+  local EXTRA="${@}"
+  # add the VLAN to the tap
+  bridge vlan add dev ${INTERFACE} vid ${VID} ${EXTRA}
+  # add the VLAN to the bridge uplink
+  bridge vlan add dev ${BRIDGE_UPLINK} vid ${VID}
+}
+
+function setup_bridge_vlan_aware {
+  BRIDGE_UPLINK="$(bridge link show | grep "master ${LINK} state" | grep -v tap | cut -d' ' -f2)"
+  local VLANS=( ${VLAN//:/ } )
+  local NUM_VLANS=${#VLANS[@]}
+  local FIRST_VID="${VLANS[0]}"
+  # only one VLAN supplied -> access port: VLAN=123 or VLAN=.123
+  if [[ ${NUM_VLANS} -eq 1 ]]; then
+    bridge_vlan_add ${FIRST_VID#.} pvid untagged
```

Comment:
You might still want a single VLAN to be tagged for some reason. With that in mind, there's no reason to special-case NUM_VLANS == 1, all you need is to check whether the first (and possibly only) VLAN is tagged or not.